From 5f5d42b390cbca6ee4dc3a0db21e5e6406b38d93 Mon Sep 17 00:00:00 2001
From: Dave Airlie <airlied@redhat.com>
Date: Fri, 12 Sep 2014 01:56:47 -0400
Subject: [drm] i915: check connector->encoder before using it

Message-id: <1410487019-24661-10-git-send-email-airlied@redhat.com>
Patchwork-id: 93919
O-Subject: [RHEL7.1 BZ1140440 PATCH 09/21] drm/i915: check connector->encoder before using it.
Bugzilla: 1140440
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Rob Clark <rclark@redhat.com>

bugzilla: 1140440
upstream: 36cd7444c044806cd2a4e450a8385597221d5d25

DP MST will need connectors that aren't connected to specific
encoders, add some checks in advance to avoid oopses.

Signed-off-by: Dave Airlie <airlied@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/i915_irq.c b/drivers/gpu/drm/i915/i915_irq.c
index d7e5593..fcc0bdc 100644
--- a/drivers/gpu/drm/i915/i915_irq.c
+++ b/drivers/gpu/drm/i915/i915_irq.c
@@ -1169,6 +1169,8 @@ static void i915_hotplug_work_func(struct work_struct *work)
  dev_priv->hpd_event_bits = 0;
  list_for_each_entry(connector, &mode_config->connector_list, head) {
   intel_connector = to_intel_connector(connector);
+  if (!intel_connector->encoder)
+   continue;
   intel_encoder = intel_connector->encoder;
   if (intel_encoder->hpd_pin > HPD_NONE &&
       dev_priv->hpd_stats[intel_encoder->hpd_pin].hpd_mark == HPD_MARK_DISABLED &&
@@ -1199,6 +1201,8 @@ static void i915_hotplug_work_func(struct work_struct *work)
 
  list_for_each_entry(connector, &mode_config->connector_list, head) {
   intel_connector = to_intel_connector(connector);
+  if (!intel_connector->encoder)
+   continue;
   intel_encoder = intel_connector->encoder;
   if (hpd_event_bits & (1 << intel_encoder->hpd_pin)) {
    if (intel_encoder->hot_plug)
diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index f729307..baf1818 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -4943,20 +4943,23 @@ static void intel_connector_check_state(struct intel_connector *connector)
        "wrong connector dpms state\n");
   WARN(connector->base.encoder != &encoder->base,
        "active connector not linked to encoder\n");
-  WARN(!encoder->connectors_active,
-       "encoder->connectors_active not set\n");
 
-  encoder_enabled = encoder->get_hw_state(encoder, &pipe);
-  WARN(!encoder_enabled, "encoder not enabled\n");
-  if (WARN_ON(!encoder->base.crtc))
-   return;
+  if (encoder) {
+   WARN(!encoder->connectors_active,
+        "encoder->connectors_active not set\n");
+
+   encoder_enabled = encoder->get_hw_state(encoder, &pipe);
+   WARN(!encoder_enabled, "encoder not enabled\n");
+   if (WARN_ON(!encoder->base.crtc))
+    return;
 
-  crtc = encoder->base.crtc;
+   crtc = encoder->base.crtc;
 
-  WARN(!crtc->enabled, "crtc not enabled\n");
-  WARN(!to_intel_crtc(crtc)->active, "crtc not active\n");
-  WARN(pipe != to_intel_crtc(crtc)->pipe,
-       "encoder active on the wrong pipe\n");
+   WARN(!crtc->enabled, "crtc not enabled\n");
+   WARN(!to_intel_crtc(crtc)->active, "crtc not active\n");
+   WARN(pipe != to_intel_crtc(crtc)->pipe,
+        "encoder active on the wrong pipe\n");
+  }
  }
 }
 
-- 
1.7.1